diff --git a/node_modules/ganache-core/index.js b/node_modules/ganache-core/index.js
index 9bdb70e..e785cc7 100644
--- a/node_modules/ganache-core/index.js
+++ b/node_modules/ganache-core/index.js
@@ -1,6 +1,4 @@
 // make sourcemaps work!
-require("source-map-support/register");
-
 const debug = require("debug")("ganache");
 
 // we use optional dependencies which may, or may not exist, so try native first
diff --git a/node_modules/ganache-core/lib/provider.js b/node_modules/ganache-core/lib/provider.js
index 5b7ac21..990fe4f 100644
--- a/node_modules/ganache-core/lib/provider.js
+++ b/node_modules/ganache-core/lib/provider.js
@@ -1,6 +1,4 @@
 // make sourcemaps work!
-require("source-map-support/register");
-
 let ProviderEngine = require("web3-provider-engine");
 let SubscriptionSubprovider = require("web3-provider-engine/subproviders/subscriptions");
 
diff --git a/node_modules/ganache-core/lib/server.js b/node_modules/ganache-core/lib/server.js
index 83bd172..404ec07 100644
--- a/node_modules/ganache-core/lib/server.js
+++ b/node_modules/ganache-core/lib/server.js
@@ -1,6 +1,4 @@
 // make sourcemaps work!
-require("source-map-support/register");
-
 var Provider = require("./provider");
 var webSocketServer = require("./webSocketServer");
 var httpServer = require("./httpServer");
diff --git a/node_modules/ganache-core/public-exports.js b/node_modules/ganache-core/public-exports.js
index 5b2c85e..7f054a0 100644
--- a/node_modules/ganache-core/public-exports.js
+++ b/node_modules/ganache-core/public-exports.js
@@ -1,6 +1,4 @@
 // make sourcemaps work!
-require("source-map-support/register");
-
 const Provider = require("./lib/provider");
 const Server = require("./lib/server");
 
