diff --git a/node_modules/@0x/sol-compiler/lib/src/utils/compiler.js b/node_modules/@0x/sol-compiler/lib/src/utils/compiler.js
index 3f22564..22afa17 100644
--- a/node_modules/@0x/sol-compiler/lib/src/utils/compiler.js
+++ b/node_modules/@0x/sol-compiler/lib/src/utils/compiler.js
@@ -210,7 +210,7 @@ function getSolcJSReleasesAsync(isOfflineMode) {
                         return [2 /*return*/, constants_1.constants.SOLC_BIN_PATHS];
                     }
                     if (!(solcJSReleasesCache === undefined)) return [3 /*break*/, 3];
-                    return [4 /*yield*/, fetch('https://ethereum.github.io/solc-bin/bin/list.json')];
+                    return [4 /*yield*/, fetch('https://raw.githubusercontent.com/ethereum/solc-bin/gh-pages/bin/list.json')];
                 case 1:
                     versionList = _a.sent();
                     return [4 /*yield*/, versionList.json()];
@@ -235,7 +235,7 @@ function compileSolcJSAsync(solcInstance, standardInput) {
         var standardInputStr, standardOutputStr, compiled;
         return __generator(this, function (_a) {
             standardInputStr = JSON.stringify(standardInput);
-            standardOutputStr = solcInstance.compileStandardWrapper(standardInputStr);
+            standardOutputStr = solcInstance.compile(standardInputStr);
             compiled = JSON.parse(standardOutputStr);
             return [2 /*return*/, compiled];
         });
@@ -252,7 +252,7 @@ function compileDockerAsync(solcVersion, standardInput) {
         var standardInputStr, dockerCommand, standardOutputStr, compiled;
         return __generator(this, function (_a) {
             standardInputStr = JSON.stringify(standardInput, null, 2);
-            dockerCommand = "docker run -i -a stdin -a stdout -a stderr ethereum/solc:" + solcVersion + " solc --standard-json";
+            dockerCommand = "docker run -i -a stdin -a stdout -a stderr ethereum/solc:" + solcVersion + " --standard-json";
             standardOutputStr = child_process_1.execSync(dockerCommand, { input: standardInputStr }).toString();
             compiled = JSON.parse(standardOutputStr);
             return [2 /*return*/, compiled];
diff --git a/node_modules/@0x/sol-compiler/lib/src/utils/constants.js b/node_modules/@0x/sol-compiler/lib/src/utils/constants.js
index 4c7aca0..26b6ac2 100644
--- a/node_modules/@0x/sol-compiler/lib/src/utils/constants.js
+++ b/node_modules/@0x/sol-compiler/lib/src/utils/constants.js
@@ -3,7 +3,7 @@ Object.defineProperty(exports, "__esModule", { value: true });
 var path = require("path");
 exports.constants = {
     SOLIDITY_FILE_EXTENSION: '.sol',
-    BASE_COMPILER_URL: 'https://ethereum.github.io/solc-bin/bin/',
+    BASE_COMPILER_URL: 'https://raw.githubusercontent.com/ethereum/solc-bin/gh-pages/bin/',
     LATEST_ARTIFACT_VERSION: '2.0.0',
     SOLC_BIN_DIR: path.join(__dirname, '..', '..', '..', 'solc_bin'),
     SOLC_BIN_PATHS: {
